Support for large entropy length

RonEld opened the pull request:
## Description
Loop on the entropy `len` until the whole output buffer is filled.
Don't return an error if `len` > `MBEDTLS_ENTROPY_BLOCK_SIZE`.
Fixes #770



## Status
**READY**

## Requires Backporting

Yes
mbedtls-2.1
mbedtls-1.3

## Migrations
NO

## Additional comments
fixes an issue where only `MBEDTLS_SHA256_C` is defined, and the desired entropy length in hmac_drbg is larger than `MBEDTLS_ENTROPY_BLOCK_SIZE`

## Todos
- [ ] Tests
- [x] Documentation
- [x] Changelog updated
- [ ] Backported


## Steps to test or reproduce
* set `MBEDTLS_SHA256_C` only in *config.h*
* use `mbedtls_entropy_func` as the entropy function callback, for `mbedtls_hmac_drbg_seed`
* use hmac_drbg as RNG

andresag01 commented on library/entropy.c:
```diff
@@ -285,87 +285,88 @@ int mbedtls_entropy_gather( mbedtls_entropy_context *ctx )
-    if( ( ret = mbedtls_mutex_lock( &ctx->mutex ) ) != 0 )
-        return( ret );
+        if( ( ret = mbedtls_mutex_lock( &ctx->mutex ) ) != 0 )
+            return(ret);
```
Style: Please put space before and after brackets.

RonEld replied:
ok

andresag01 commented on library/entropy.c:
```diff
@@ -285,87 +285,88 @@ int mbedtls_entropy_gather( mbedtls_entropy_context *ctx )
+        if( ctx->initial_entropy_run == 0 )
+        {
+            ctx->initial_entropy_run = 1;
+            if( ( ret = mbedtls_entropy_update_nv_seed( ctx ) ) != 0 )
```
I am not sure about this, do we really want to update the nv_seed for very large polls?

RonEld replied:
This API was to define an NV seed according to the desired size of the seed. I think it's valid

andresag01 commented on library/entropy.c:
```diff
@@ -285,87 +285,88 @@ int mbedtls_entropy_gather( mbedtls_entropy_context *ctx )
+            ctx->source[i].size = 0;
 
-    memcpy( output, buf, len );
+        memcpy( output+ ( j * MBEDTLS_ENTROPY_BLOCK_SIZE), buf, len > MBEDTLS_ENTROPY_BLOCK_SIZE ? MBEDTLS_ENTROPY_BLOCK_SIZE : len );
```
Style: Please break this line so that it does not exceed 79 characters

RonEld replied:
ok

andresag01 commented on library/entropy.c:
```diff
@@ -285,87 +285,88 @@ int mbedtls_entropy_gather( mbedtls_entropy_context *ctx )
-    memcpy( output, buf, len );
+        memcpy( output+ ( j * MBEDTLS_ENTROPY_BLOCK_SIZE), buf, len > MBEDTLS_ENTROPY_BLOCK_SIZE ? MBEDTLS_ENTROPY_BLOCK_SIZE : len );
+        j++;
+        len -= len > MBEDTLS_ENTROPY_BLOCK_SIZE ? MBEDTLS_ENTROPY_BLOCK_SIZE : len;
```
Style: Please break this line so that it does not exceed 79 characters long.

RonEld replied:
ok

andresag01 commented on library/entropy.c:
```diff
@@ -285,87 +285,88 @@ int mbedtls_entropy_gather( mbedtls_entropy_context *ctx )
+            for( i = 0; i < ctx->source_count; i++ )
+                if( ctx->source[i].size < ctx->source[i].threshold )
+                    done = 0;
+        } while (!done);
```
Style: Please put space before and after brackets.

RonEld replied:
ok

andresag01 commented on library/entropy.c:
```diff
@@ -285,87 +285,88 @@ int mbedtls_entropy_gather( mbedtls_entropy_context *ctx )
+        {
+            ctx->initial_entropy_run = 1;
+            if( ( ret = mbedtls_entropy_update_nv_seed( ctx ) ) != 0 )
+                return(ret);
```
Style: Please put space before and after brackets.

RonEld replied:
ok

andresag01 commented on ChangeLog:
```diff
@@ -41,6 +41,9 @@ Bugfix
      constructed certificates to bypass the certificate verification check.
+   * Loop on length to seed, until all desired length is achieved, not to return
+     an error when the desired entropy length is higher than
+     MBEDTLS_ENTROPY_BLOCK_SIZE. Fix for #770.
```
I think we should be a bit more specific in the ChangeLog, for example by mentioning the function that changed.

RonEld replied:
ok

andresag01 commented on library/entropy.c:
```diff
@@ -285,87 +285,88 @@ int mbedtls_entropy_gather( mbedtls_entropy_context *ctx )
 
 int mbedtls_entropy_func( void *data, unsigned char *output, size_t len )
```
Please note that this PR changes the behaviour of the function, but it does not update the doxygen documentation for `mbedtls_entropy_func()` that still says:
```
 * \param len       Number of bytes desired, must be at most MBEDTLS_ENTROPY_BLOCK_SIZE
```

RonEld commented:
Hi @andresag01 Thanks for your comments. I agree that it's not ideal to change the entropy source function, but I think that we shoujd support seed requests larger than `MBEDTLS_ENTROPY_BLOCK_SIZE`
The configuration is valid, in my opinion, and thus we get into the error raised in the issue mentioned n the description
As for your suggestions:
1.  I think that `mbedtls_entropy_func`is more of an example, since it is used in all our documentations, without mentioning it's an example. Having said that, I DO think we should have some configuration for `MBEDTLS_ENTROPY_SOURCE_ALT` or something similar. Telling a user to give a different entropy pool function that can handle 48 bytes is basically covering our issue, IMHO.
2. I think modifying HMAC_DRBG is prefered, but as the comment in the code mentions:
~~~
    /*
     * See SP800-57 5.6.1 (p. 65-66) for the security strength provided by
     * each hash function, then according to SP800-90A rev1 10.1 table 2,
     * min_entropy_len (in bits) is security_strength.
     *
     * (This also matches the sizes used in the NIST test vectors.)
~~~
So I think that we cannot reduce the size of the entropy length for SHA256, or we should force `MBEDTLS_ENTROPY_SHA512_ACCUMULATOR` for entropy , meaning require `MBEDTLS_SHA512_C` when using `MBEDTLS_ENTROPY_C` , even if the MD is SHA256. I think this is not really an option.

mpg commented:
One thing is not clear to me: in case some user provides their own entropy function (which our API allows), would they need to change their entropy function? I think changing our own entropy function is fine as long as the permissible arguments are a superset of what they used to be, but imposing that change on user's functions raises more questions IMO.

andresag01 commented:
@RonEld:
>I think modifying HMAC_DRBG is prefered, but as the comment in the code mentions

I did not mean to reduce the size of the entropy required by HMAC_DRBG, instead I meant that we should modify the function `mbedtls_hmac_drbg_reseed()` so that it calls `f_entropy()` as many times as necessary until it has the full 48 (or whatever) bytes of entropy it wants. I am *definitely* against reducing the entropy intake for HMAC_DRBG.

In summary, what I was trying to say with my review comment is: I am not sure whether we should change the behaviour for `mbedtls_entropy_func()` so that it returns an arbitrary number of entropy bytes OR we should take care of that somewhere else. I think I prefer the latter option because it does not change the behaviour of `mbedtls_entropy_func()` at all, but also its not like this change is going to break anyone so I can be easily convinced either way.

RonEld commented:
Hi @andresag01 I understand your point and I agree with you. I''ll refactor.
However, I think that we shouldn't limit our entropy function to `MBEDTLS_ENTROPY_BLOCK_SIZE`, since as I see it, the user should define what's the entropy length they wish

andresag01 commented:
@RonEld: Just to clarify, I was just giving a few suggestions and to be honest I am not sure what is the most desirable option at this stage. I think we should ask for a third opinion...

RonEld commented:
I understand, but I agree with you that for this specific scenario, it would be cleaner to modify the HMAC_DRBG code.
But, IMHO, our entropy function should allow any kind of entropy length

sbutcher-arm commented:
@RonEld - This PR needs rework. There's unresolved review feedback, plus it looks like it needs rebasing.

RonEld commented:
I have rebased to resolve conflicts.
I have also fixed the style errors raised in the review.

RonEld commented:
I added an iteration in `mbedtls_hmac_drbg_reseed()` to collect as many times as needed the entropy, as @andresag01 suggested, as it is cleaner for the *current use case*.
However, I kept the iteration in `mbedtls_entropy_func()`, as I think our entropy function should also allow returning as much entropy as requested.

yanesca commented:
My thoughts on this PR:
- We had the unwritten requirement that entropy functions have to be able to return MBEDTLS_ENTROPY_BLOCK_SIZE bytes of entropy. Now we are changing this to the requirement of entropy functions having to return arbitrary amount of entropy. This might break users who implemented their own entropy function.
- Also, fixing our own entropy function leaves users with custom entropy functions that can't return more than MBEDTLS_ENTROPY_BLOCK_SIZE bytes vulnerable.
- Our entropy function is properly documented, trying to get more than MBEDTLS_ENTROPY_BLOCK_SIZE bytes of entropy is violation of the API.
- There might be room for improvement in our entropy module interfaces, but I am not sure that this is the right way to do it. Also I don't think that we should do it as part of a bugfix PR unless it is a clear improvement compared to other options.

Therefore I think:
- Whatever we decide to do on development we should not backport the entropy function change, it can break users and potentially leaves users vulnerable.
- If we decide to change the entropy function on development we should take extra care to explain the security implications in the changelog.
- If we decide to change the entropy function I would prefer to do it in a more readable way, that function is getting pretty complicated. (For example moving the old functionality in a static helper function and have the public function do the extra layer of polling.)

RonEld commented:
very well, I shall revert the changes in the entropy function, and keep the changes in hmac_drbg alone.

I will add similar functionality to ctr_drbg, with tests

RonEld commented:
1. I have reverted the changes in the entropy function, keeping only the fix in `hmac_drbg`.
2. I have added documentation in our entropy function prototype that `len` should not exceed `MBEDTLS_ENTROPY_BLOCK_SIZE `. As mentioned by @yanesca , this was an unwritten requirement, but since it is unwritten, I am not sure we aren't breaking anything for existing users.
3. I have modified the test entropy functions to follow that requirement.
4. I have fixed ctr_drbg reseed function to follow same logic as hmac_drbg.

yanesca commented on include/mbedtls/entropy.h:
```diff
@@ -95,7 +95,7 @@ extern "C" {
  * \param data      Callback-specific data pointer
  * \param output    Data to fill
- * \param len       Maximum size to provide
+ * \param len       Maximum size to provide, must be at most MBEDTLS_ENTROPY_BLOCK_SIZE.
```
I have several concerns here:

1. This is the entropy callback used by the entropy pool and this documentation shouldn't be changed. 

The callback we need to document is in `hmac_drbg.h` and `ctr_drbg.h` in the context and as a parameter to the corresponding seed function.

2. Also, since this is a documentation of a callback that people will implement, I think it would be best to
- Make it clear if we are imposing requirements on the caller or the implementer
- Present it from the implementers point of view, as a requirement on the implementation

I suggest something like:
> \warning \p f_entropy MUST fill the buffer with full entropy and MUST be able to do so for buffer lengths up to at least `MBEDTLS_ENTROPY_BLOCK_SIZE` bytes.

3. Even if we don't change our entropy function here and now I agree that it is the right thing moving forward. Getting there would be much smoother if we would require implementors to handle arbitrary requests (even if our implementation does not do it yet), therefore I suggest that we do so on development and keep the milder requirement on the backport branches.

RonEld replied:
>1. This is the entropy callback used by the entropy pool and this documentation shouldn't be changed.

I disagree. Since this entropy function is used by our drbg functions, we expect implementors this callback accept `MBEDTLS_ENTROPY_BLOCK_SIZE`  at most.

>1. Also, since this is a documentation of a callback that people will implement, I think it would be best to
> * Make it clear if we are imposing requirements on the caller or the implementer
> * Present it from the implementers point of view, as a requirement on the implementation

This is why I think that the warning should be in the entropy callback description

yanesca replied:
> I disagree. Since this entropy function is used by our drbg functions, we expect implementors this callback accept `MBEDTLS_ENTROPY_BLOCK_SIZE` at most.

No, this function is used by our entropy pool and not our drbg functions:
https://github.com/ARMmbed/mbedtls/blob/9c99dc862cd5cf9cdc62cbfcd200063c83295397/library/entropy.c#L143-L145

Here is the function used by our DRBG functions:
https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/ctr_drbg.h#L143
https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/ctr_drbg.h#L183
https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/hmac_drbg.h#L98
https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/hmac_drbg.h#L142

RonEld replied:
Thanks for the correction, for some reason I confused the entropy source function callback with the entropy function callback. I'll fix that

yanesca commented on ChangeLog:
```diff
@@ -19,6 +19,8 @@ Bugfix
    * Reduce stack usage of `mpi_write_hlp()` by eliminating recursion.
      Fixes #2190.
+   * Assure that required seed length does not exceed MBEDTLS_ENTROPY_BLOCK_SIZE,
+     by looping in thereseed functions of ctr_drbg and hmac_drbg. Fixes #770.
```
typo: thereseed -> the reseed

yanesca commented on ChangeLog:
```diff
@@ -19,6 +19,8 @@ Bugfix
      in X.509 module. Fixes #2212.
    * Reduce stack usage of `mpi_write_hlp()` by eliminating recursion.
      Fixes #2190.
+   * Assure that required seed length does not exceed MBEDTLS_ENTROPY_BLOCK_SIZE,
```
This is not what we do. In the code we are making sure that seeding CTR DRBG and HMAC DRBG is possible even if the randomness source can't return more than MBEDTLS_ENTROPY_BLOCK_SIZE at a time. Given that this is a requirement we want to lift in the near future, I suggest we focus on the actual issue fixed and don't go into details. I was thinking something like this:

> Fix issue in seeding HMAC DRBG when MBEDTLS_SHA512_C is disabled. Reported by dbedev. Fixes #770.

RonEld commented:
@yanesca I addressed your comments.
Please Review.
I will squash the latest commit to https://github.com/ARMmbed/mbedtls/pull/1063/commits/5e05607b20033d14ad22daecf7ef052c166184fb and https://github.com/ARMmbed/mbedtls/pull/1063/commits/e4ef7abffa6152a34ddcb814c9ee288600509c54 once approved

RonEld commented:
@yanesca Thank you for reviewing!

I have cleared the history a bit. Please reapprove
I will backport the fix to `mbedtls-2.7` and `mbedtls-2.16` shortly.
@andresag01 please review

RonEld commented:
Backports raised in #2456 and #2457

yanesca commented:
@RonEld You seem to have changed the bases and did the cleanup as well. In the future can you please do two separate force pushes, one for the cleanup and one for changing bases so that we can review easier?

RonEld commented:
@yanesca Thank you for your review and comment.
AFAIK, the way to cleanup the log is by interactive rebasing.
Is there another way?

yanesca commented:
@RonEld Yes, that is the way how I do it too. But you don't have to switch bases while doing that. If you do the interactive rebase on a commit that is already on your branch, then you can tidy up your history and avoid any other change. If you force push after doing so, then the reviewer can see with a single click what you have done.

RonEld commented:
Both backports have been fully reviewed and approved, so removing the "needs backports" label

cc @mpg @Patater

gilles-peskine-arm commented on tests/suites/test_suite_hmac_drbg.misc.data:
```diff
@@ -10,6 +10,10 @@ HMAC_DRBG entropy usage SHA-256
 hmac_drbg_entropy_usage:MBEDTLS_MD_SHA256
 
+HMAC_DRBG entropy usage SHA-256 accumulator
+depends_on:MBEDTLS_SHA256_C:!MBEDTLS_SHA512_C
```
Nowadays the entropy accumulator also uses SHA-256 if `MBEDTLS_ENTROPY_FORCE_SHA256` is set.

gilles-peskine-arm commented:
I'm closing this PR because it now needs to be made against Mbed Crypto. Since the main issue that it fixes (https://github.com/ARMmbed/mbedtls/issues/770) has been already resolved (https://github.com/ARMmbed/mbed-crypto/pull/238), there is no urgency.

RonEld commented:
We should consider what to do with this PR's backports though.
Should we close them or not?

gilles-peskine-arm commented:
I think this PR no longer needs backports. The main part (HMAC_DRBG doesn't work at all when used in the normal way and the entropy module uses SHA-256) has been resolved. The rest (DRBG modules may not work when using an entropy function that returns less output than the library-provided one) is a limitation, not an outright bug, and so should not be backported.
